ruby_programming/testing_with_rspec/test_driven_development.md

```diff
@@ -0,0 +1,37 @@
+# Test-Driven Development
+
+## Introduction
+
+Hopefully, you've been having fun developing in Ruby thus far. Perhaps one thing you may **not** have enjoyed as much, though, is having to run your script on the command line, running through each step manually, only to find that it doesn't return what you expected. And then comes the *even-less-fun* part: You tug at your hair a bit, change a few variables, and run your script again through each step, **hoping** that the results are correct.
+
+Folks, we're here to tell you there *IS* better. Hoping is for that exam you didn't study for and took by the seat of your pants. **Knowing** is test-driven development.
+
+## Learning Outcomes
+
+- Understand what test-driven development (TDD) is, and why it's important.
+
+## What is Test-Driven Development?
+
+[Test-Driven Development (TDD)](https://en.wikipedia.org/wiki/Test-driven_development) is a process and technique of software development that relies on the repetition of a very short and specific development cycle. In each cycle, requirements (i.e., what you want your code to do) are turned into specific test cases first. These requirements could be anything from an entire feature that requires end-to-end (E2E) testing, such as a tests that cover a user logging into your website succesfully AND unsuccessfully, to a new Ruby class you've devised, for which unit tests might suffice. Either way, the test suite for these requirements fail initially, since actual code hasn't been written yet. 
+
+Once the code is written and passes our test suite, you can move onto refactoring. As you might have guessed, running the test suite against the changes we make to our features or units enables us to refactor *while* making sure everything still works.
+
+Colloquially, this process is often referred to as the "**red-green-refactor**" cycle. That's all it is, in a nutshell: automated tests **drive** the design of software. You don't need to know how the entire architecture of your sweet, new, industry-breaking application will work. Your application only has to be broken down, step-by-step, until small units are identified and covered by tests.
+
+## Why Is It Important?
+
+Even if TDD intuitively seems like good practice, its utility is hotly debated even today. 
+
+No one is saying that tests aren't useful. Testing at [all levels](https://robots.thoughtbot.com/rails-test-types-and-the-testing-pyramid) is crucial. Testing ensures that your application is working as intended. Testing provides a safety net that gives your application room to grow and evolve as painlessly as possible as features are added and dependencies change.
+
+The real question is whether or not tests should *drive* development, or whether or not it's important for them to come *first*. Many developers write tests after they've written their code. There is definitely a longer list of pros than cons for having tests in general—no one will disagree with that—so should we test first? 
```

Comment:
There is definitely a longer list of pros than cons for having tests in general, no one will disagree with that. So should we test first?